RTClib.cpp

```diff
@@ -746,6 +746,168 @@ TimeSpan::TimeSpan(int16_t days, int8_t hours, int8_t minutes, int8_t seconds)
 /**************************************************************************/
 TimeSpan::TimeSpan(const TimeSpan &copy) : _seconds(copy._seconds) {}
 
+/*!
+    @brief  Create a new TimeSpan from an iso8601 formatted period string
+
+    If the string is entirely malformed (doesn't begin with P or -) this will
+    construct a TimeSpan of 0 seconds. If parsing fails before the end of the
+    string, this constuctor will interpret its argument as the longest
+    well-formed prefix string.
+
+    For example, the string P5M10~ will parse the same as P5M
+
+    @param iso8601 the formatted string
+*/
+TimeSpan::TimeSpan(char *iso8601) : _seconds(0) {
+  int32_t sign = 1;
+  char *cursor = iso8601;
+  if (*cursor == '-') {
+    sign = -1;
+    cursor++;
+  }
+  if (*cursor == 'P') {
+    cursor++;
+  } else {
+    return;
+  }
+
+  char *rest;
+  long num = strtol(cursor, &rest, 10);
+  cursor = rest;
+  if (*cursor == 'D') {
+    _seconds += sign * SECONDS_PER_DAY * num;
+    cursor++;
+  }
+
+  if (*cursor == 'T') {
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  } else if (*cursor == '\0') {
+    return;
+  } else {
+    // error: malformed iso8601
+    return;
+  }
+
+  if (*cursor == 'H') {
+    _seconds += sign * SECS_PER_MIN * MINS_PER_HOUR * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor == 'M') {
+    _seconds += sign * SECS_PER_MIN * num;
+    cursor++;
+    num = strtol(cursor, &rest, 10);
+    cursor = rest;
+  }
+
+  if (*cursor == 'S') {
+    _seconds += sign * num;
+  }
+}
+
+/*!
+    @brief  Formats this TimeSpan according to iso8601
+
+    See toCharArray for more details on the format
+
+    @return String of formatted TimeSpan
+*/
+String TimeSpan::toString() const {
+  constexpr size_t buflen = 19;
+  char buf[buflen];
+  this->toCharArray(buf, buflen);
+  return String(buf);
+}
+
+/*!
+    @brief  Formats this TimeSpan according to iso8601
+
+    Fails (returning 0) if buf is too small to store the result. buf size 19
+    will never fail.
+    Example: TimeSpan t(32, 23, 54, 11) formats to "P32DT23H54M11S".
+
+    @param buf char array to write output. Always contains trailing '\0'.
+    @param len the length of buf. Must be at least 1.
+    @return number of bytes written excluding trailing '\0' or 0 on failure
+*/
+size_t TimeSpan::toCharArray(char *buf, size_t len) const {
+  size_t written = 0;
+
+  if (_seconds == 0) {
+    written = snprintf(buf, len, "PT0S");
+    if (written >= len) {
+      buf[len - 1] = '\0';
+      return 0;
+    } else {
+      return written;
+    }
+  }
+
+  int8_t sign = _seconds > 0 ? 1 : -1;
+  int32_t tmp = _seconds;
+  int8_t seconds = sign * (tmp % SECS_PER_MIN);
+  tmp /= SECS_PER_MIN;
```

Comment:
The sign could be folded into `tmp` at this point, in order to avoid extra multiplications later.
```suggestion
  tmp = sign * (tmp / MINS_PER_HOUR);
```
Maybe add a comment explaining that it is not folded earlier in order to avoid the overflow on `INT32_MIN`.

Reply:
It could be folded in. I decided to avoid that optimization to keep it readable.

Reply:
I folded it in.